File: Solutions/PE007_ten_thousand_first_prime/python/ten_thousand_first_prime.py
#!/bin/python3

# https://www.hackerrank.com/contests/projecteuler/challenges/euler007


def sieve_of_eratosthenes(n):
    """
    Return list of primes less than n
    """
    res = [2]
    i = 3
    marked = set()
    while i <= n**.5:
        if i not in marked:
            res.append(i)
            j = 0
            while j <= n/i:
                marked.add(i + j*i)
                j += 1
        i += 2
    while i <= n:
        if i not in marked:
            res.append(i)
        i += 2
    return res


if __name__ == '__main__':
    first_10000_primes = sieve_of_eratosthenes(104729)
    t = int(input().strip())
    for a0 in range(t):
        n = int(input().strip())
        print(first_10000_primes[n-1])

File: Solutions/PE007_ten_thousand_first_prime/cpp/TenThousandFirstPrime.cpp
#include <iostream>
#include <vector>
#include <set>
#include <math.h>

//https://www.hackerrank.com/contests/projecteuler/challenges/euler007

using namespace std;

vector<long> SieveOfEratosthenese(int n) {
	// Return list of primes less than n
	vector<long> res;
	res.push_back(2);
	int i = 3;
	set<long> marked;
	while (i <= sqrt(n)) {
		if (marked.find(i) == marked.end()) {
			res.push_back(i);
			long j = 0;
			while (j <= n / i) {
				marked.insert(i + j * i);
				j++;
			}
		}
		i += 2;
	}
	while (i <= n) {
		if (marked.find(i) == marked.end())
			res.push_back(i);
		i += 2;
	}
	return res;
}

int main() {
	vector<long> First10000Primes = SieveOfEratosthenese(104729);
	int t;
	cin >> t;
	for (int i = 0; i < t; i++) {
		int n;
		cin >> n;
		cout << First10000Primes[n-1] << endl;
	}
}
